src/QUICConnection.ts

```diff
@@ -162,159 +146,115 @@ class QUICConnection extends EventTarget {
-    const abortProm = promise<never>();
-    const abortHandler = () => {
-      abortProm.rejectP(ctx.signal.reason);
-    };
-    ctx.signal.addEventListener('abort', abortHandler);
-    const connection = new this(args);
-    // If it's a server connection we want to pass the initial packet
-    const data = args.type === 'server' ? args.data : undefined;
-    // This ensures that TLS has been established and verified on both sides
-    try {
-      await Promise.race([
-        Promise.all([
-          connection.start(data),
-          connection.establishedP,
-          connection.secureEstablishedP,
-        ]),
-        abortProm.p,
-      ]);
-    } catch (e) {
-      const code =
-        args.reasonToCode != null ? (await args.reasonToCode(e)) ?? 0 : 0;
-      await connection.stop({
-        applicationError: false,
-        errorCode: code,
-        errorMessage: e.message,
-        force: true,
-      });
-      throw e;
-    } finally {
-      ctx.signal.removeEventListener('abort', abortHandler);
-    }
-    connection.logger.debug('secureEstablishedP');
-    // After this is done
-    // We need to establish the keep alive interval time
-    if (connection.config.keepAliveIntervalTime != null) {
-      connection.startKeepAliveIntervalTimer(
-        connection.config.keepAliveIntervalTime,
-      );
+  /**
+   * This event represents the fact that `this.conn.close()` has already been called.
+   * It does not actually mean that `closedP` is resolved.
+   * That only occurs if the timeout timer detects that the connection is closed.
+   */
+  protected handleEventQUICConnectionClose = async (evt: events.EventQUICConnectionClose) => {
+    // The final send is necessary after a close
+    // But even if this fails, then we expect a timeout expiry to occur
+    // Failure of send is both a caller and domain error
+    // We skip the caller error, the domain error is also somewhat ignored
+    // Because this handler is registered to only run once
+    // It will proceed to stop the connection
+
+    // Only run this if the close wasn't due to peer
+    if (evt.detail.type !== 'peer') {
+      await this.send();
     }
 
-    return connection;
+    if (this[running]) {
+      // Here we keep forcing true, since force affects streams
+      // Not the connection itself
+      // Failing to stop is also a caller error, there's no domain error handling
+      // So we let it bubble up
+      await this.stop({ force: true, });
+    }
   }
 
+  /**
+   * Whenever there is a send event on a quic stream
+   * We will asynchronously proceed with a `this.send` call
+   * This also allows us to deal with failures here if it happens
+   */
+  protected handleEventQUICStreamSend = async () => {
+    // Failure of send is both a caller error and domain error
+    // In this case we can ignore the caller error, since the domain error will be handled
+    // by the QUICConnection error handlers
+    await this.send();
+  };
```

Comment:
Very weird error here. the `await this.send()` call is throwing `ErrorQUICConnectionNotRunning`.  It's weird since the connection has not actually stopped yet.

I can see in the logs that the connection stops RIGHT after this error. So it's likely a race condition there where the connection is stopping before logging out the stopping message.

The weird thing is, the `send()` method is allowed to run when stopping. So maybe there's a bug with `async-init` that this is not working properly?

Reply:
Can you confirm this?

Reply:
Will this require a check for the connection is running too? Again same issue because handling could be async.

Reply:
Fixed in fedb61c0df3dab7be0a4c72f3d7206b6d5d3a778. One diff is that `QUICConnection.send` can be done while it is stopping, but `QUICSocket.send` cannot be done if it is stopping.

Reply:
This has been resolved by `@matrixai/async-init` version `^1.10.0`.